pkg/queue/queue.go

```diff
@@ -0,0 +1,117 @@
+package queue
+
+import (
+	"container/list"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Queue provides a queue for type T.
+// The queue will allow to push, pull, and pull blocking of items.
+// Additionally, it exposes safety guards such as a max size as well as metrics to track the queue growth and size.
+type Queue[T comparable] struct {
+	maxSize        int
+	counterMetric  *prometheus.CounterVec
+	queue          *list.List
+	notEmptySignal concurrency.Signal
+	mutex          sync.Mutex
+}
+
+// OptionFunc provides options for the queue.
+// Note that the comparable type is currently required, once we upgrade to go1.20 we can switch this to
+// any and creation will be much easier.
+type OptionFunc[T comparable] func(queue *Queue[T])
+
+// WithCounterVec provides a counter vec which tracks added and removed items from the queue.
+func WithCounterVec[T comparable](vec *prometheus.CounterVec) OptionFunc[T] {
+	return func(queue *Queue[T]) {
+		queue.counterMetric = vec
+	}
+}
+
+// WithMaxSize provides a limit to the size of the queue. By default, no size limit is set so the queue is
+// unbounded.
+func WithMaxSize[T comparable](size int) OptionFunc[T] {
+	return func(queue *Queue[T]) {
+		queue.maxSize = size
+	}
+}
+
+// NewQueue creates a new queue. Optionally, a metric can be included.
+func NewQueue[T comparable](opts ...OptionFunc[T]) *Queue[T] {
+	queue := &Queue[T]{
+		notEmptySignal: concurrency.NewSignal(),
+		queue:          list.New(),
+	}
+
+	for _, opt := range opts {
+		opt(queue)
+	}
+
+	return queue
+}
+
+// Pull will pull an item from the queue. If the queue is empty, the default value of T will be returned.
+// Note that his does not wait for items to be available in the queue, use PullBlocking instead.
+func (q *Queue[T]) Pull() T {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if q.queue.Len() == 0 {
+		var nilT T
+		return nilT
+	}
+
+	item := q.queue.Remove(q.queue.Front()).(T)
+
+	if q.counterMetric != nil {
+		q.counterMetric.With(prometheus.Labels{"Operation": metrics.Remove.String()}).Inc()
+	}
+
+	if q.queue.Len() == 0 {
+		q.notEmptySignal.Reset()
+	}
+
+	return item
+}
+
+// PullBlocking will pull an item from the queue, potentially waiting until one is available.
+// In case the waitable signals done, the default value of T will be returned.
+func (q *Queue[T]) PullBlocking(waitable concurrency.Waitable) T {
+	var item T
+	for item == *new(T) {
```

Comment:
Why do we need to wrap `select` in `for` here?

Reply:
So just from a pure Go PoV without the outer `for` it'd work, however for concurrency safety:

2 Go routines are running `PullBlocking`. A single item gets added into the queue. Both go routines receive the signal and call `Pull`. One of them acquires the lock, while the other waits. Whichever go routine acquired the lock, will receive the newly added item. If there's no item added once the second go routine acquired the lock, the returned item will be nil / default value, and `PullBlocking` would return nil / default value, which would break it's contract / purpose.

With the additional `for` around, go routines will wait for the next signal in case of multiple consumers and ensure that _eventually_ they'll get an non-default value item.